### ISSUE TITLE
Method 'need' must be implemented by CompUnit::Repository

### ISSUE THREAD

**Issue by hythm7** [rakudo/rakudo#4949]
## The Problem

started getting the below error after [this](https://github.com/rakudo/rakudo/commit/50025c1a48dd194696d275acadb9c7fd5b82f9a2) commit when installing modules that contain `use CompUnit::Repository::Staging`

```
Method 'need' must be implemented by CompUnit::Repository because it is required by roles: CompUnit::Repository.
```

## Expected Behavior
No Error.

## Actual Behavior
Error.

## Steps to Reproduce
Create a module structure like  this and run `t/test.raku`:

```
.
├── lib
│   └── MyModule.rakumod
├── META6.json
└── t
    └── test.raku

> cat lib/MyModule.rakumod 
use CompUnit::Repository::Staging;

> cat t/test.raku 
use CompUnit::Repository::Staging;

CompUnit::Repository::Staging.new(:prefix</tmp/stage>, :name<stage>, next-repo => $*REPO).install(Distribution::Path.new('/tmp/delme/MyModule'.IO));

> cat META6.json 
{
  "perl" : "6.*",
  "name" : "MyModule",
  "version" : "0.0.0",
  "description" : "test",
  "depends" : [],
  "test-depends": [],
  "auth": "github:hythm7-o",
  "provides" : {
    "MyModule": "lib/MyModule.rakumod"
  },
  "source-url" : ""
}

```

## Environment
* Operating system: Linux
* Compiler version (`perl6 -v` or `raku -v`):
```
Welcome to Rakudo™ v2022.04-16-g50025c1a4.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2022.04.
```

**Comment by lizmat** [comment]
Question: why would `MyModule.rakumod` need to have a `use CompUnit::Repository::Staging;` ??

**Comment by hythm7** [comment]
It is needed in the full version because `CompUnit::Repository::Staging` is used by `class` `method`s

**Comment by lizmat** [comment]
Is `'/tmp/delme/MyModule'` referring to the directory with the `META6.json` ?

**Comment by hythm7** [comment]
> Is `'/tmp/delme/MyModule'` referring to the directory with the `META6.json` ?

Yes.

**Comment by hythm7** [comment]
Any chance we can revert the mentioned commit before the coming release?  It has been broken since 2022.04 and I need to update my local rakudo installation but i cant before this issue is addressed.

**Comment by lizmat** [comment]
I will look at this again in the coming days.  Pretty sure we'll be able to fix this before the next release.

**Comment by lizmat** [comment]
I cannot spend enough time on this before the 2022.07 release, so I've reverted the commit in question with https://github.com/rakudo/rakudo/commit/4fc3d455b0 .

Please close this issue if confirmed this fixes it.

**Comment by melezhik** [comment]
passed test in r3 - http://sparrowhub.io:2222/report/927

**Comment by hythm7** [comment]
Thanks. I don't have time to test now, but closing as per quoted comment

> passed test in r3 - http://sparrowhub.io:2222/report/927

**Comment by melezhik** [comment]
fail for me now on the latest commit - https://gist.github.com/melezhik/5d5998fe69d1504af5fc141b8a19f13e

**Comment by hythm7** [comment]
Reopening because [this line](https://github.com/rakudo/rakudo/commit/50025c1a48dd194696d275acadb9c7fd5b82f9a2#diff-d6c2d6f915688a3ac33dc4f8706c92fbe2679b195b0d8d1e66e38c4e26c0bcecL7) also introduces another issue. not registering `CUR::Staging` name in `CU::RepositoryRegistry` makes some distributions fails to compile (looks for `dist` file in the wrong prefix).


Error: 
```sh
===SORRY!=== Error while compiling /tmp/workdir/MyModule/home#sources/EFBD61D0EFF20D6053C529EEEC7FEC6803EC62D9 (MyModule)
An exception occurred while evaluating a constant
at /tmp/workdir/MyModule/home#sources/EFBD61D0EFF20D6053C529EEEC7FEC6803EC62D9 (MyModule):9
Exception details:
  Failed to open file /home/hythm/.raku/dist/0A9FF6017CA75303FC6271D4E739E7D0C7AC83FB: No such file or directory
    in sub guess_library_name at /home/hythm/.rakubrew/versions/moar-blead/install/share/perl6/core/sources/C9715CAFA1689E465708C242DCC4397969C49295 (NativeCall) line 218
    in method setup at /home/hythm/.rakubrew/versions/moar-blead/install/share/perl6/core/sources/C9715CAFA1689E465708C242DCC4397969C49295 (NativeCall) line 305
    in method setup at /home/hythm/.rakubrew/versions/moar-blead/install/share/perl6/core/sources/C9715CAFA1689E465708C242DCC4397969C49295 (NativeCall) line 366
    in sub raku-nativecall at /home/hythm/.rakubrew/versions/moar-blead/install/share/perl6/core/sources/16BC91A643D1DDF8224FC60D8D3251689EF308AC (NativeCall::Dispatcher) line 46
    in block  at /home/hythm/.raku/sources/EFBD61D0EFF20D6053C529EEEC7FEC6803EC62D9 (MyModule) line 9


  in block <unit> at t/test.raku line 10
```

To reproduce:

```sh
mkdir /tmp/workdir;
mkdir /tmp/workdir/MyModule;
cd /tmp/workdir/MyModule/;
mkdir lib resources t;
touch resources/resource;

cat << 'EOF' > lib/MyModule.rakumod
unit class MyModule;

use NativeCall;

constant $c = %?RESOURCES<resource>;

sub s() is native($c) { }

constant C = s();
EOF

cat << 'EOF' > t/test.raku
use CompUnit::Repository::Staging;

my $dist = Distribution::Path.new('.'.IO);

my $stage := CompUnit::Repository::Staging.new:
    prefix    => '/tmp/workdir/stage'.IO,
    name      => 'home',
    next-repo => $*REPO;

$stage.install: $dist;
EOF

cat << 'EOF' > META6.json
{"meta-version":1,"perl":"6.c","depends":[ ],"auth":"cpan:author","description":"","license":"Artistic-2.0","name":"MyModule","provides":{"MyModule":"lib/MyModule.rakumod"},"authors":[ ],"version":"0.60","resources":["resource"],"source-url":"","api":"0"}
EOF

raku t/test.raku
```

**Comment by hythm7** [comment]
Regarding the first error `Method 'need' must be implemented by CompUnit::Repository` which is triggered by `CUR::Staging candidates()` when staging for custom repo (e.g `/opt/my-app-repo`), although not common as staging for default Raku repos (e.g. home site vendor core) but still I think it is a valid use case and shoul not be left broken in curren  `CUR::Staging`.

**Comment by hythm7** [comment]
Fixed with 454752e3e3b20c48724263851874efea68a54adb

**Comment by melezhik** [comment]
confirming the fix on sparkyci
